src/scripts/exclusion_term_expansion.py

```diff
@@ -0,0 +1,378 @@
+        search_result_df: pd.DataFrame = sparql_jinja2_robot_query(
+            onto_path=onto_path,
+            query_template_path=SPARQL_TERM_REGEX_PATH,
+            regexp=regex_pattern,
+            cache_suffix=f'r{str(index)}')
+        if len(search_result_df) > 0:
+            search_result_df['term_id'] = search_result_df['term_id'].apply(uri_to_curie)
+            # todo later: #x1: would be good to filter out by IRI or prefix that is not from source ontology
+            #  ...This can be done (a) in Python, or (b) in SPARQL, if using that.
+            # search_result_df = search_result_df[search_result_df['term_id'].str.startswith(prefix)]
+            for fld in exclusion_fields:
+                search_result_df[fld] = row[fld]
+            search_results.append(search_result_df)
+    if search_results:
+        # - Combine each of the regex search results
+        df_regex_pre_results = pd.concat(search_results)
+        # - Query for children
+        df_regex_results = expand_ontological_exclusions(onto_path, df_regex_pre_results, cache_suffix='2')
+
+    # Combine & massage
+    # - CONCAT: Combine results of all queries
+    df_results = pd.concat([x for x in [df_explicit_results, df_regex_results] if x is not None])
+    # - Drop duplicates: As there is likely to be overlap between (a) explicitly listed terms, and (b) Regex-selected
+    # terms. It is also theoretically possible that there may be some cases where there is a different exclusion_reason
+    # for a term that is within both (a) and (b). In that case, duplicate terms will remain, showing both reasons.
+    df_results = df_results.drop_duplicates()
+    # TODO: put above into a func: expand_intensional_exclusions()
+    # TODO: Put below into a func: get_non_inclusions()
+    # TODO Non-inclusions are terms that:
+    #   in: reports/mirror_signature-omim.tsv
+    #   not in: reports/component_signature-omim.tsv
+    #   have one of prefixes
+    # TODO: understand and program reason why we need the above and not relevant_sig:
+    #  ...
+    # - Filter: terms not in relevant_signature.txt, e.g. terms actually in onto that we care about
+    # TODO: print any that were intensional exclusion expansions but showed in relevant sig
+    #  ...and add as a comment fo nico
+    df_relevant_terms = pd.read_csv(relevant_signature_path).fillna('')
+    df_relevant_terms['term'] = df_relevant_terms['term'].apply(uri_to_curie)
+    relevant_terms: List[str] = list(set(df_relevant_terms['term']))
+    # todo: some redundancy in these 2 lines that could be reduced. see if i can assign ~df_results... to a variable
+    #   and re-use that for both the print statement and the df_results2 filter
+    relevant_terms_accidentally_excluded = [x for x in relevant_terms if x in set(df_results['term_id'])]
+    if relevant_terms_accidentally_excluded:
+        print(f'Warning: The extensionalization of intensional exclusion table resulted in the following terms which '
```

Comment:
I realized that this edge case could happen, and decided to handle it and print a warning if it occurs:

```python
print(f'Warning: The extensionalization of intensional exclusion table resulted in the following terms which '
              f'would have been excluded, but are contained within {os.path.basename(relevant_signature_path)}, and'
              f' thus have been removed from exclusion: {relevant_terms_accidentally_excluded}', file=sys.stderr)
```

Reply:
@matentzn @sabrinatoro @nicolevasilevsky Hmm, interesting. Glad I put this warning here. I was running ICD10CM and I did notice that there are terms in the `relevant_signature` file that appear in the `exclusions` file:

```
python /exclusion_term_expansion.py -n icd10cm
Warning: The extensionalization of intensional exclusion table resulted in the following terms which would have been excluded, but are contained within icd10cm_relevant_signature.txt, and thus have been removed from exclusion: ['ICD10CM:E89-E89', 'ICD10CM:P96.3', 'ICD10CM:P92', 'ICD10CM:N99-N99', 'ICD10CM:P94.8', 'ICD10CM:R00-R09', 'ICD10CM:P05-P08', 'ICD10CM:Z19-Z19', 'ICD10CM:P91.8', 'ICD10CM:R10-R19', 'ICD10CM:Z66-Z66', 'ICD10CM:P94', 'ICD10CM:D78-D78', 'ICD10CM:P94.0', 'ICD10CM:P00-P04', 'ICD10CM:O85-O92', 'ICD10CM:P96.5', 'ICD10CM:E16.3', 'ICD10CM:O60-O77', 'ICD10CM:C7B-C7B', 'ICD10CM:P52', 'ICD10CM:K40-K46', 'ICD10CM:P36', 'ICD10CM:P96.9', 'ICD10CM:Z14', 'ICD10CM:M40', 'ICD10CM:P96.8', 'ICD10CM:M96-M96', 'ICD10CM:P91.61', 'ICD10CM:P96.89', 'ICD10CM:P91', 'ICD10CM:E36-E36', 'ICD10CM:P92.0', 'ICD10CM:P92.2', 'ICD10CM:E16.0', 'ICD10CM:P94.2', 'ICD10CM:V10-V19', 'ICD10CM:P91.63', 'ICD10CM:P92.3', 'ICD10CM:V70-V79', 'ICD10CM:R47-R49', 'ICD10CM:E16.2', 'ICD10CM:T79-T79', 'ICD10CM:X52-X58', 'ICD10CM:P93', 'ICD10CM:E15', 'ICD10CM:Z40-Z53', 'ICD10CM:X10-X19', 'ICD10CM:P92.5', 'ICD10CM:P91.2', 'ICD10CM:P96.0', 'ICD10CM:Z30-Z39', 'ICD10CM:P91.62', 'ICD10CM:V80-V89', 'ICD10CM:P91.4', 'ICD10CM:Y90-Y99', 'ICD10CM:P60', 'ICD10CM:C7A-C7A', 'ICD10CM:P51', 'ICD10CM:P91.0', 'ICD10CM:P91.811', 'ICD10CM:P91.6', 'ICD10CM:X71-X83', 'ICD10CM:Z00-Z13', 'ICD10CM:Z77-Z99', 'ICD10CM:P91.821', 'ICD10CM:R30-R39', 'ICD10CM:P09-P09', 'ICD10CM:P93.8', 'ICD10CM:Y62-Y69', 'ICD10CM:R20-R23', 'ICD10CM:R25-R29', 'ICD10CM:P96', 'ICD10CM:M26', 'ICD10CM:P96.82', 'ICD10CM:P92.8', 'ICD10CM:R97-R97', 'ICD10CM:P59', 'ICD10CM:B95-B97', 'ICD10CM:V00-Y99', 'ICD10CM:P92.09', 'ICD10CM:V50-V59', 'ICD10CM:V40-V49', 'ICD10CM:P50', 'ICD10CM:Z55-Z65', 'ICD10CM:V00-V09', 'ICD10CM:P91.88', 'ICD10CM:H59-H59', 'ICD10CM:P84-P84', 'ICD10CM:P92.4', 'ICD10CM:V90-V94', 'ICD10CM:Y83-Y84', 'ICD10CM:Z69-Z76', 'ICD10CM:R70-R79', 'ICD10CM:P91.82', 'ICD10CM:P96.81', 'ICD10CM:Z67-Z67', 'ICD10CM:W00-W19', 'ICD10CM:P91.829', 'ICD10CM:P93.0', 'ICD10CM:P92.6', 'ICD10CM:P92.01', 'ICD10CM:P96.1', 'ICD10CM:R90-R94', 'ICD10CM:V30-V39', 'ICD10CM:X92-Y09', 'ICD10CM:V98-V99', 'ICD10CM:P91.60', 'ICD10CM:O09-O09', 'ICD10CM:H95-H95', 'ICD10CM:M99-M99', 'ICD10CM:P91.822', 'ICD10CM:O00-O08', 'ICD10CM:Z18-Z18', 'ICD10CM:Y21-Y33', 'ICD10CM:V60-V69', 'ICD10CM:P96.83', 'ICD10CM:P95', 'ICD10CM:Z20-Z29', 'ICD10CM:R83-R89', 'ICD10CM:T66-T78', 'ICD10CM:P96.2', 'ICD10CM:W85-W99', 'ICD10CM:R99-R99', 'ICD10CM:P58', 'ICD10CM:V20-V29', 'ICD10CM:Y35-Y38', 'ICD10CM:P53', 'ICD10CM:P91.5', 'ICD10CM:P94.9', 'ICD10CM:V95-V97', 'ICD10CM:L76-L76', 'ICD10CM:Z68-Z68', 'ICD10CM:W65-W74', 'ICD10CM:P91.81', 'ICD10CM:R50-R69', 'ICD10CM:P90', 'ICD10CM:P92.1', 'ICD10CM:O80-O82', 'ICD10CM:X50-X50', 'ICD10CM:R00-R99', 'ICD10CM:X30-X39', 'ICD10CM:Z00-Z99', 'ICD10CM:P91.1', 'ICD10CM:R80-R82', 'ICD10CM:P10-P15', 'ICD10CM:P54', 'ICD10CM:P91.9', 'ICD10CM:O30-O48', 'ICD10CM:T15-T19', 'ICD10CM:X00-X08', 'ICD10CM:P91.3', 'ICD10CM:E16.1', 'ICD10CM:Y70-Y82', 'ICD10CM:W50-W64', 'ICD10CM:P92.9', 'ICD10CM:P94.1', 'ICD10CM:Z17-Z17', 'ICD10CM:P91.819', 'ICD10CM:R40-R46', 'ICD10CM:L49', 'ICD10CM:E16.4', 'ICD10CM:P91.823', 'ICD10CM:P80-P83', 'ICD10CM:W20-W49', 'ICD10CM:J95-J95']
```

I was thinking that maybe this is not a real problem? Maybe it the case that the `relevant_signature` file is an initial selection of relevant terms, but the exclusions are meant to override `relevant_signature`? But I think not, because Nico asked me to explicitly not exclude terms that are in `relevant_signature`.

So I'm curious as to why this is happening. Perhaps it is an artefact of the complexities of multiple inheritance from polyhierarchy? Or perhaps these terms just should not have been added to the exclusion table?

Reply:
@matentzn Prior to last Friday, I was instructed to use the `relevant_signature` file. The idea was to look at the extensionalized list of exclusions and filter out any that were in `relevant_signature`. However, I was just now getting all the warnings that I crossed out above, and 0 'relevant' extensionalized exclusions in my results.

Then I looked at your [google doc](https://docs.google.com/document/d/1-bFd6GeHcLHImjKImML_j0uAyKENIlb5NN64oGP4rwY/edit?disco=AAAAc-gOiSs), and noticed that `relevant_signature` was dropped from the requirements. As such, I've discarded the use for `relevant_signature`. Just want to get final verification from you that this is indeed the intent; that our exclusion tables are essentially overriding `relevant_signature`.

Reply:
Replaced by e.g. `reports/component_signature-omim.tsv`

Reply:
Alright, thanks for that confirmation. Removing everything related to `relevant_signature` from the PR now.